File: reference/flutter-src/lib/domain/stats/log_filters.dart
import '../models/detection.dart';

/// Date-range filter for the history log.
enum DateRange { all, week }

/// The two history filters combined. `speciesId == null` (or `'all'`) means all
/// species.
class LogFilter {
  const LogFilter({this.speciesId, this.range = DateRange.all});

  final String? speciesId;
  final DateRange range;

  bool get isAllSpecies => speciesId == null || speciesId == 'all';

  LogFilter copyWith({String? speciesId, DateRange? range}) => LogFilter(
        speciesId: speciesId ?? this.speciesId,
        range: range ?? this.range,
      );
}

/// Apply [filter] to [detections]. Pure and deterministic — [now] is injected
/// so the "this week" window is testable. Order is preserved.
List<Detection> applyFilters(
  List<Detection> detections,
  LogFilter filter, {
  required DateTime now,
}) {
  final cutoff = now.subtract(const Duration(days: 7));
  return detections.where((d) {
    final speciesOk = filter.isAllSpecies || d.speciesId == filter.speciesId;
    final rangeOk =
        filter.range == DateRange.all || d.timestamp.isAfter(cutoff);
    return speciesOk && rangeOk;
  }).toList();
}

File: app/src/main/kotlin/com/mozzid/domain/stats/LogFilters.kt
package com.mozzid.domain.stats

import com.mozzid.domain.model.Detection

/** Date-range filter for the history log. */
enum class DateRange { ALL, WEEK }

/**
 * The two history filters combined. [speciesId] == null (or "all") means all
 * species.
 */
data class LogFilter(
    val speciesId: String? = null,
    val range: DateRange = DateRange.ALL,
) {
    val isAllSpecies: Boolean get() = speciesId == null || speciesId == "all"
}

/**
 * Apply [filter] to [detections]. Pure and deterministic — [nowMillis] is injected
 * so the "this week" window is testable. Order is preserved.
 */
fun applyFilters(
    detections: List<Detection>,
    filter: LogFilter,
    nowMillis: Long,
): List<Detection> {
    val cutoff = nowMillis - 7L * 24 * 60 * 60 * 1000
    return detections.filter { d ->
        val speciesOk = filter.isAllSpecies || d.speciesId == filter.speciesId
        val rangeOk = filter.range == DateRange.ALL || d.timestampMillis > cutoff
        speciesOk && rangeOk
    }
}
